src/build/build_step.go

```diff
@@ -464,7 +464,19 @@ func calculateAndCheckRuleHash(state *core.BuildState, target *core.BuildTarget)
 	// Set appropriate permissions on outputs
 	if target.IsBinary {
 		for _, output := range target.FullOutputs() {
-			if err := os.Chmod(output, target.OutMode()); err != nil {
+			// Walk through the output,
+			// if the output is a directory,apply output mode to the file instead of the directory
+			err := fs.Walk(output, func(path string, isDir bool) error {
+				if isDir && path == output {
```

Comment:
don't think this is right? surely you should never chmod it if `isDir`?

would suggest something like
```
if isDir {
    return nil
}
return os.Chmod(path, target.OutMode())
```

Reply:
That's true, it's likely we'd have more nested dir as well